### PR TITLE
Rebase of --serve commits onto v1.16

### PR DESCRIPTION
**Pull request by r3pwn**
This is a rebase of the "--serve commits" ( #1419 ) onto the latest master branch, with all merge issues resolved.

Still seems to be a bit buggy, but this at least puts it on the most recent master branch cleanly

### CODE REVIEW

**Comment by r3pwn** [comment]
I'm using @rom1v's [suggestion of testing using VLC](https://github.com/Genymobile/scrcpy/issues/1159#issuecomment-589282647) and it does appear to work for a bit. I had it crash on me once, but haven't been able to reproduce. Additionally, the device window is also visible, which I would classify as a bug. Looking into it

**Comment by rom1v** [comment]
> Rebase of …

Good idea :+1:

> Merge branch 'master' of github.com:Genymobile/scrcpy

You should never merge a branch in a PR (just rebase instead) :wink:

**Comment by r3pwn** [comment]
@rom1v My bad, I have resolved that and properly rebased it

**Comment by r3pwn** [comment]
After some more testing, I haven't been able to reproduce that crash I initially saw, and I realized if I want the display to not appear, it's possible to pass in `--no-display`

**Comment by r3pwn** [comment]
@rom1v Have you had the chance to take a look at this PR?

File: app/src/cli.c [hunk 1]
```diff
@@ -626,6 +632,169 @@ guess_record_format(const char *filename) {
...
+        LOGE("Unexpected protocol (expected tcp)");
+        return false;
+    }
+    
+    //Check if the choosen ip is valid
+    if (!strcmp(ip, "localhost")) {
+        ip_value = 0x7F000001;
+        ip_valid = true;
+    } else if (check_if_ip_valid(ip)) {
+        ip_value = convert_ip_to_int(ip);
```
Comment (intergratedsolutions) [inline app/src/cli.c:774]: I'm trying to find my employees IP Addresses.

File: app/src/cli.c [hunk 2]
```diff
@@ -860,8 +1038,8 @@ scrcpy_parse_args(struct scrcpy_cli_args *args, int argc, char *argv[]) {
         }
     }
 
-    if (!opts->display && !opts->record_filename) {
-        LOGE("-N/--no-display requires screen recording (-r/--record)");
+    if (!opts->display && !opts->record_filename && !opts->serve) {
+        LOGE("-N/--no-display requires screen recording (-r/--record) or to serve to another client (--serve)");
```
Comment (intergratedsolutions) [inline app/src/cli.c:1042]: I need screen recordings

**Review by intergratedsolutions — APPROVED** [review]

**Comment by intergratedsolutions** [comment]
> This is a rebase of the "--serve commits" ( #1419 ) onto the latest master branch, with all merge issues resolved.
> 
> Still seems to be a bit buggy, but this at least puts it on the most recent master branch cleanly

Thankyou

**Comment by rom1v** [comment]
Please read https://github.com/Genymobile/scrcpy/pull/1419#issuecomment-1013964650

**Pull request closed without merging**